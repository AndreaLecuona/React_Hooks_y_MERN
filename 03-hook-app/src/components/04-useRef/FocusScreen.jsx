import React, { useRef } from 'react';
import '../02-useEffect/effects.css';

//usos comunes de useRef -> mantener una referencia mutable

//permite cambiar la referencia sin necesidad de volver a renderizar 
export const FocusScreen = () => {

    const inputRef = useRef();

    const handleClick = () => {
        // document.querySelector('input').focus(); o...
        //document.querySelector('input').select(); que además del focus en el borde, también seleciona el texto
    
        // USO POCO COMUN PERO POSIBLE
        //reemplazo el querySelector anterior con:
        inputRef.current.select();
        console.log(inputRef);
    }

    return (
        <div>
            <h1>FocusScreen</h1>
            <hr/>

            <input
            ref={ inputRef }
            className='form-control'
            type="text"
            placeholder='Tu nombre'
            />

            <button className="btn btn-outline-primary mt-4"
            onClick={ handleClick }
            >
                Focus
            </button>
        </div>
    )
}
