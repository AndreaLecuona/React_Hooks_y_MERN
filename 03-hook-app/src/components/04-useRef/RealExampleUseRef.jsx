import React, { useState } from 'react';
import '../02-useEffect/effects.css';
import { MultipleCustomHooks } from '../03-examples/MultipleCustomHooks';

export const RealExampleUseRef = () => {

    const [show, setShow] = useState(false);

    return (
        <div>
            <h1>Caso de uso real del useRef</h1>
            <hr/>

            {/* el useRef nos va a ayudar en casos de error, cuando el componente tarda en cargar los datos de la api y el usuario lo esconde antes que finalice su tarea de cargar las frases */}
            {/* lo vamos a aplicar en el custom hook useFetch */}
            
            {show && <MultipleCustomHooks/>}

            <button className="bt btn-primary mt-5"
            onClick={ () => {
                setShow( !show );
            }}>
                Show/Hide
            </button>
        </div>
    )
}
