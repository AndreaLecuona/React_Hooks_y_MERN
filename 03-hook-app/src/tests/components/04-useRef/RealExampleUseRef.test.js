import { shallow } from "enzyme"
import { RealExampleUseRef } from "../../../components/04-useRef/RealExampleUseRef"

describe( 'Pruebas en RealExampleUseRef', () => { 
    
    // lo coloco aca porque no va a modificarse durante las pruebas
    const wrapper = shallow( <RealExampleUseRef/> ); // si tuviera argumentos hay que ponerlos

    test('debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'MultipleCustomHooks' ).exists() ).toBe( false );

    })
    
    test('debe mostrar el componente MultipleCustomHooks', () => {
        // en el componente cuando se hace click, se modifica un state show a true y eso hace que se renderice y visibilice el componente MultipleCustomHooks
        wrapper.find( 'button' ).simulate( 'click' );

        expect( wrapper.find( 'MultipleCustomHooks' ).exists() ).toBe( true );

    })

})