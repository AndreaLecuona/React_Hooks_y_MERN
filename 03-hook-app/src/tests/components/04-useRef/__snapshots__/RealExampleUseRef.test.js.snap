// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RealExampleUseRef debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Caso de uso real del useRef
  </h1>
  <hr />
  <button
    className="bt btn-primary mt-5"
    onClick={[Function]}
  >
    Show/Hide
  </button>
</div>
`;
